### ISSUE TITLE
Support llvm source-based coverage in the Engine unittests

### ISSUE THREAD

**Issue by gw280** [flutter/flutter#86385]
It'd be great to be able to support source-based coverage reporting for our unit test suites. LLVM supports this and documentation is at https://clang.llvm.org/docs/SourceBasedCodeCoverage.html

This requires the following:

~~- Add the coverage GN configs to buildroot~~
~~- Update engine gn script to add a coverage flag to enable the config~~
~~- Apply the coverage GN config to Engine files being built~~
~~- Add a coverage script that generates the coverage reports~~

- Modify the existing run-tests script to optionally output the coverage profile data when running the test suites
- Modify CI recipes to call the coverage script and upload the reports somewhere

**Comment by gw280** [comment]
cc @zanderso @chinmaygarde @CaseyHillers

**Comment by chinmaygarde** [comment]
I had already wired this up in https://github.com/flutter/engine/pull/7576. At that time, our version of clang was too old. This is no longer an issue. To demonstrate usage, [I had uploaded a sample of the report](https://public.chinmaygarde.com/flutter/coverage/2019-10-16/coverage/index.html) after manually updating my version of clang. We should wire that up on CI now.

**Comment by CaseyHillers** [comment]
Where do we plan to upload the code coverage to? Are we planning to use an existing service, or build our own?

**Comment by gw280** [comment]
It looks like it's only wired up on mac or gcc right now?

**Comment by chinmaygarde** [comment]
The gist of it is, use the `--coverage` on `./flutter/tools/gn` and use `generate_coverage.py` specifying the unit-tests to run and merge coverage reports of.

> Where do we plan to upload the code coverage to?

I don't think we have planned anything. Whatever we can push to from the bots.

> It looks like it's only wired up on mac or gcc right now?

It should work for any Clang based toolchains. I had initially wired it up on Mac. On Linux, there should be no blockers as long as the toolchain definitions there understand the `--coverage` gn flag. It ought to be trivial. AFAIk, its only a couple of additional flags on each driver invocation.

**Comment by chinmaygarde** [comment]
`gcc_toolchain.gni` (which the clang toolchain definitions also use) already [has this wired up](https://github.com/flutter/buildroot/blob/5c61df1b8e5ae701c1bd76f089b888b51326a72a/build/toolchain/gcc_toolchain.gni#L117). Can you run a sample right now? I bet it JustWorks™.

**Comment by gw280** [comment]
For now let's just dump the generated HTML coverage reports on a google cloud storage bucket and set it up for web access - https://cloud.google.com/storage/docs/hosting-static-website

**Comment by chinmaygarde** [comment]
cc @gw280 Were you able to test if if the existing options were sufficient to get you the coverage reports you were after?

**Comment by gw280** [comment]
Yep, it looks like it. We just need to figure out running it on CI now.
